### DISCUSSION TITLE
VS2022: AttributeError: module '_pytest.compat' has no attribute '_translate_non_printable'

### DISCUSSION THREAD

**Discussion by ejm-revvity** [pytest-dev/pytest#12222] [General] [answered]
Is this a pytest bug or a Visual Studio 2022 bug?
[https://github.com/microsoft/PTVS/issues/7844](https://github.com/microsoft/PTVS/issues/7844)

Reverting to pytest v8.0.2 fixes the issue.
I realize the 8.1.0 got yanked but 8.1.1 does not fix it either i.e. it reproduces the same issue.

**Comment by RonnyPfannschmidt** [comment]
vscode bug, thats a internal api

**Reply by ejm-revvity** [reply]
VS2022 means Visual Studio 2022, not VS Code 2022 - just to be clear - I updated above.
And yes, this is breaking an internal VS 2022 api in the pytest test runner.

But whose responsibility is it to fix this?
pytest 8.0.2 works and does not break the VS 2022 test runner.
However, 8.1.0 and 8.1.1 do? Is it a Microsoft Visual Studio support responsibility to modify the test runner to be compatible with pytest 8.1.1?
That seems strange.

**Reply by The-Compiler** [reply]
If they use API that's private, it's indeed their responsibility to keep things up to date if said private API changes - or ideally, they wouldn't be using private API at all. What's strange about it?

**Reply by ejm-revvity** [reply]
I agree 100%. I guess what is strange is the apparent misunderstanding due to language, maybe?

What has changed is in fact pytest. pytest has changed from version 8.0.2 to version 8.1.0 and now to version 8.1.1. Using Visual Studio version 17.9.6 with version 8.0.2 of pytest works and presents no known issues. However, using the exact same version of Visual Studio, 17,9.6, with version 8.1.1 of pytest, does break the VS test runner. The 'private API' for VS test runner has not changed. Only pytest has changed.

**Reply by The-Compiler** [reply, marked as answer]
I don't know how else to put this. Private/internal API (prefixed with `_` in Python, i.e. `_pytest`) is exactly that: It's private and internal, and might change at any time. Yes, pytest changed, but things broke because VS is using things which aren't meant to be used by external projects in the first place. Thus, it's their job to either fix that, or adjust to pytest's changes.

**Reply by ejm-revvity** [reply]
Ah - so indeed I misunderstood to which 'private API' you were referring.

You are claiming that that the VS test runner is using/accessing a private field within a pytest class or module (or something akin to that) and that private field either no longer exists or has changed name. Therefore, the VS support team should either adjust appropriately or better yet, refactor to not access the pytest private API at all.

Thank you. I think I can file a bug now as I have a vested interest in getting this fixed.
